_posts/2021-06-08-quarkus-deploys-apps-on-openshift.adoc

```diff
@@ -0,0 +1,170 @@
+
+== First attempts
+
+Given that complexity we created our own tool for configuration and deployment. The initial version was based on Ansible and ran into Tower workflows. For different reasons we decided to rewrite it in Java a year ago, mainly to:
+
+- Get better performances
+- Use a high level language that could cope with algorithm complexity
+- Get better support for unit and integration testing
+- Improve productivity with quality development tools (e.g. IDE)
+
+One thing we were concerned about though, was memory consumption, CPU usage and startup times since for a given application, each tenant is deployed in isolation in its own deployment process. For that reason we decided to not depend on any application framework when we started the Java rewrite, because if they do offer productivity and standardization, the abstraction they provide comes with a resource cost. So the program was written in plain Java, and it was small enough in size and use cases that we were able get away with it, provided we relied on a few patterns (e.g. constructor based injection by hand) for clean code. This program is called _ocpdeploy_.
+
+As we were moving away from Tower, we decided to leverage our Kubernetes infrastructure by running _ocpdeploy_ as tasks in Tekton pipelines. This gave us repeatability, a way to deploy different applications, or even different tenants for the same application with different _ocpdeploy_ versions. The use of java (any other high level language would have been a good fit) provided us with a high level of productivity, and maintainable code, while being able to implement complex algorithms for configuration processing, and raised the level of quality of _ocpdeploy_ releases thanks to our extensive regression testing suite.
+
+All was well... until we started deploying multiple tenants at the same time, and/or different applications. We had sized the _ocpdeploy_ container to `200 millicores` and `280 Mb` of RAM. For some applications we have around 30 tenants, that we all deploy in parallel. This meant 30 pods, which accounted for `6 cores` and `8 Gb` of RAM. This seems livable, but deployments tend to be done after business hours, between 7 and 8pm for many. And we started being afraid of the impact _ocpdeploy_ itself would have if we were running multiple deployments in parallel, on top of our SpringBoot or WildFly applications, which take their toll on the cluster at startup.
+
+== Enter the Quarkus Universe
+
+So less than a year ago we decided to launch a Quarkus POC on _ocpdeploy_, the goal being to deploy it as a GraalVM executable.
+The main challenges we ran into were related to the lack of support for GraalVM in some of the libraries we were using: FreeMarker, SMBJ, the Fabric8 OpenShift client (when instantiating OpenShift specific CRDs), and a home made java agent that ships logs to a proprietary centralized logging system.
```

Comment:
spell out "Server Message Block in Java" the first time possibly link to it?

Reply:
I do not understand your comment. what is "Server Message Block in Java"?

Reply:
first hit on google when i looked up smbj - https://github.com/hierynomus/smbj

if that is not what it is even more reason to have it spelled out :)

Reply:
that one is funny. I did not know smb was standing for "server message block". I thought it meant samba. but actually smb is the protocol, and samba a product. and smb does stand for "server message block", not samba. so yes, I will expain.